Mathlib/CategoryTheory/Shift/Adjunction.lean

```diff
@@ -0,0 +1,458 @@
+  slice_lhs 3 4 => rw [← Functor.map_comp, ← Functor.map_comp, Iso.inv_hom_id_app]
+  simp only [Functor.comp_obj, Functor.map_id, id_comp, assoc]
+
+/--
+The commutation isomorphisms of `Adjunction.RightAdjointCommShift.iso` are compatible with
+the unit of the adjunction.
+-/
+lemma iso_compat_unit (a : A) : CommShift.compat_unit adj a (F.commShiftIso a) (iso adj a) := by
+  intro X
+  rw [← cancel_mono ((RightAdjointCommShift.iso adj a).inv.app (F.obj X)), assoc, assoc,
+    Iso.hom_inv_id_app]
+  dsimp
+  rw [comp_id]
+  simp only [RightAdjointCommShift.iso_inv_app adj _ _ (neg_add_cancel a), Functor.id_obj,
+    Functor.comp_obj]
+  apply (adj.homEquiv _ _).symm.injective
+  dsimp
+  simp only [homEquiv_counit, Functor.map_comp, assoc, counit_naturality, Functor.comp_obj,
+    Functor.id_obj, counit_naturality_assoc, left_triangle_components_assoc]
+  erw [← NatTrans.naturality_assoc]
+  dsimp
+  rw [shift_shiftFunctorCompIsoId_hom_app, Iso.inv_hom_id_app_assoc,
+    Functor.commShiftIso_hom_naturality_assoc, ← Functor.map_comp,
+    left_triangle_components, Functor.map_id, comp_id]
+
+end RightAdjointCommShift
+
+variable (A)
+
+/--
+Given an adjunction `F ⊣ G` and a `CommShift` structure on `F`, this constructs
+the unique compatible `CommShift` structure on `G`.
+-/
+@[simps]
+noncomputable def rightAdjointCommShift [F.CommShift A] : G.CommShift A where
+  iso a := RightAdjointCommShift.iso adj a
+  zero := by
+    refine CommShift.compat_unit_unique_right adj 0 (F.commShiftIso 0)  _ _
+      (RightAdjointCommShift.iso_compat_unit adj 0) ?_
+    rw [F.commShiftIso_zero]
+    exact CommShift.compat_unit_isoZero adj
+  add a b := by
+    refine CommShift.compat_unit_unique_right adj (a + b) (F.commShiftIso (a + b))  _ _
+      (RightAdjointCommShift.iso_compat_unit adj (a + b)) ?_
+    rw [F.commShiftIso_add]
+    exact CommShift.compat_unit_isoAdd adj a b _ _ _ _
+      (RightAdjointCommShift.iso_compat_unit adj a) (RightAdjointCommShift.iso_compat_unit adj b)
+
+lemma commShift_of_leftAdjoint [F.CommShift A] :
+    letI := adj.rightAdjointCommShift A
+    adj.CommShift A := by
+  letI := adj.rightAdjointCommShift A
+  apply CommShift.mk'
+  refine ⟨fun a => ?_⟩
+  ext X
+  dsimp
+  simp only [Functor.commShiftIso_id_hom_app, Functor.comp_obj,
+    Functor.id_obj, id_comp, Functor.commShiftIso_comp_hom_app]
+  exact RightAdjointCommShift.iso_compat_unit adj a X
+
+end Adjunction
+
+namespace Equivalence
+
+variable {C D : Type*} [Category C] [Category D] (E : C ≌ D)
+
+section
+
+variable (A : Type*) [AddMonoid A] [HasShift C A] [HasShift D A]
+
+/--
+If `E : C ≌ D` is an equivalence, this expresses the compatibility of `CommShift`
+structures on `E.functor` and `E.inverse`.
+-/
+class CommShift [E.functor.CommShift A] [E.inverse.CommShift A] : Prop where
+  commShift_unitIso_hom : NatTrans.CommShift E.unitIso.hom A := by infer_instance
+  commShift_counitIso_hom : NatTrans.CommShift E.counitIso.hom A := by infer_instance
```

Comment:
```suggestion
abbrev CommShift [E.functor.CommShift A] [E.inverse.CommShift A] : Prop :=
  E.toAdjunction.CommShift A
```
(This would make things parallel with `Adjunction.IsMonoidal` and `Equivalence.IsMonoidal` in `CategoryTheory.Monoidal.Functor`.)

Reply:
This change made the line
`attribute [instance] commShift_unitIso_hom commShift_counitIso_hom`
stop working, and I did not know what the correct substitute for it should be, so a couple proofs became more painful later. But it compiles.

Reply:
You need to add this:
```lean
variable [E.functor.CommShift A] [E.inverse.CommShift A] 

instance [E.CommShift A] : NatTrans.CommShift E.unitIso.hom A :=
  inferInstanceAs (NatTrans.CommShift E.toAdjunction.unit A)

instance [E.CommShift A]: NatTrans.CommShift E.counitIso.hom A :=
  inferInstanceAs (NatTrans.CommShift E.toAdjunction.counit A)
```

Reply:
This helped with the proof of the instance on line 423, but there are still some things that Lean was able to infer automatically in the old version and that I have to give it explicitly now (lines 395 and 406). So this does not do the same thing.